libraries/anvill_passes/src/RecoverStackFrameInformation.h

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2021 Trail of Bits, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <anvill/Lifters/Options.h>
+#include <anvill/Result.h>
+#include <llvm/Pass.h>
+
+#include <unordered_map>
+
+#include "BaseFunctionPass.h"
+
+namespace anvill {
+
+enum class StackAnalysisErrorCode {
+  UnknownError,
+  UnsupportedInstruction,
+  StackPointerResolutionError,
+  StackFrameTypeAlreadyExists,
+  InternalError,
+  InvalidParameter,
+};
+
+// Contains a list of `load` and `store` instructions that reference
+// the stack pointer
+using StackPointerRegisterUsages = std::vector<llvm::Instruction *>;
+
+// This structure contains the stack size, along with the lower and
+// higher bounds of the offsets, and all the `load` and `store` instructions
+// that have been analyzed
+struct StackFrameAnalysis final {
+  // Describes a `store` or `load` instruction that is operating on
+  // the stack, along with the SP-relative offset and the operand
+  // type
+  struct StackOperation final {
+    llvm::Instruction *instr{nullptr};
+    std::int32_t offset{};
```

Comment:
Related to https://github.com/lifting-bits/anvill/pull/131#discussion_r588736750

I have started casting offsets to std::int32_t since it seems like the std::int64_t displacement values returned by `TryResolveReference` are in fact 32-bits values that have not been sign-extended to 64-bit.

Should I try to convert to std::int64_t?

Reply:
Yeah, convert to `int64_t`, and then we should investigate `TryResolveReference` in a separate PR or Issue.